earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +190,168 @@ class khMTSet
     khLockGuard guard(mutex);
     return std::vector<T>(set.begin(), set.end());
   }
+  void clear(void) {
+    khLockGuard guard(mutex);
+    set.clear();
+  }
+  size_t size(void) {
+    return set.size();
+  }
+};
+
+
+template <class K, class V>
+class khMTMap
+{
+ protected:
+  //typedef std::unordered_map<K,V> Map;
+  typedef std::unordered_map<K,V> Map;
+  Map     map;
+  mutable std::mutex mutex;
+
+ private:
+  // private and unimplimented
+  khMTMap& operator=(const khMTMap&) = delete;
+
+  khMTMap(const khMTMap& o, const std::lock_guard<std::mutex>&) : map(o.map) { }
+  khMTMap(khMTMap&& o, const std::lock_guard<std::mutex>&) : map(std::move(o.map)) { }
+ public:
+  khMTMap(void) { }
+  khMTMap(const khMTMap& o) : khMTMap(o, std::lock_guard<std::mutex>(o.mutex)) { }
+  khMTMap(khMTMap&& o) : khMTMap(o, std::lock_guard<std::mutex>(o.mutex)){ }
+
+  void insert(const std::pair<K,V> &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.insert(val);
+  }
+  void erase(const K &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.erase(val);
+  }
+  size_t size(void) {
+    return map.size();
+  }
+  V& operator[](const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map[key];
+  }
+  void clear(void) {
+    std::lock_guard<std::mutex> guard(mutex);
+    map.clear();
+  }
+  std::vector<K> keys(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    std::vector<K> thekeys;
+    thekeys.reserve(map.size());
+    for (const auto& p : map) {
+      thekeys.push_back(p.first);
+    }
+    return thekeys;
+  }
+  bool contains(const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.find(key) != map.end();
+  }
+  bool empty(void) const {
+    return map.empty();
+  }
 };
 
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but creates some less obvious ones
+template <typename T>
+class MTVector : public std::vector<T>{
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
+
+  private:
+    mutable khRWLock mtx;
+
+    MTVector(const MTVector& a, const khReadGuard&) : Base(a) { }
+
+  public:
+    void push_back(const T& v) {
+      khWriteGuard lock(mtx);
+      Base::push_back( v );
+    }
+    MTVector() { }
+    // forward to private copy constructor to protect vector from modification
+    MTVector(const MTVector& a) : MTVector(a, khReadGuard(a.mtx)) { }
+    MTVector(const std::vector<T>& v) : Base(v) { }
+    
+
+    const T& operator[] (const size_t idx) const {
+      khReadGuard lock(mtx);
+      return Base::operator[] (idx); 
+    }
+
+    bool operator==(const MTVector& x) const {
+      // Since using a read lock, no need to worry about lock ordering or things like a==a
+      khReadGuard lock1(mtx);
+      khReadGuard lock2(x.mtx);
+      return std::equal(Base::begin(), Base::end(), x.begin());
```

Comment:
Minor: I would take out `Base::` for the `begin()` and `end()` calls here. They're not necessary, and in the (albeit unlikely) event that `begin` and `end` were overridden in MTVector, this would have unexpected results.

Reply:
It turns out that this was required for the template compilation to work with the new subclass.  I've been unable to find any workaround.